plugins/inputs/sqlserver/connectionstring.go

```diff
@@ -0,0 +1,124 @@
+)
+
+const odbcPrefix = "odbc:"
+
+const (
+	emptyServerName   = "<empty-server-name>"
+	emptyDatabaseName = "<empty-database-name>"
+)
+
+// getConnectionIdentifiers returns the sqlInstance and databaseName from the given connection string.
+// The name of the server is returned as-is in the connection string
+// If the connection string could not be parsed or sqlInstance/databaseName were not present, a placeholder value is returned
+func getConnectionIdentifiers(connectionString string) (sqlInstance string, databaseName string) {
+	if len(connectionString) == 0 {
+		return
+	}
+
+	trimmedConnectionString := strings.TrimSpace(connectionString)
+
+	if strings.HasPrefix(trimmedConnectionString, odbcPrefix) {
+		i := strings.Index(trimmedConnectionString, odbcPrefix)
+		connectionStringWithoutOdbc := trimmedConnectionString[i+len(odbcPrefix):]
+		sqlInstance, databaseName = parseConnectionStringKeyValue(connectionStringWithoutOdbc)
+	} else if strings.HasPrefix(trimmedConnectionString, "sqlserver://") {
+		sqlInstance, databaseName = parseConnectionStringURL(trimmedConnectionString)
+	} else {
+		sqlInstance, databaseName = parseConnectionStringKeyValue(trimmedConnectionString)
+	}
+
+	return
+}
+
+// parseConnectionStringKeyValue parses a "key=value;" connection string and returns the SQL instance and database name
+func parseConnectionStringKeyValue(connectionString string) (sqlInstance string, databaseName string) {
+	sqlInstance = emptyServerName
+	databaseName = emptyDatabaseName
+
+	keyValuePairs := strings.Split(connectionString, ";")
+	for _, keyValuePair := range keyValuePairs {
+		if len(keyValuePair) == 0 {
+			continue
+		}
+
+		keyAndValue := strings.SplitN(keyValuePair, "=", 2)
+		key := strings.TrimSpace(strings.ToLower(keyAndValue[0]))
+		if len(key) == 0 {
+			continue
+		}
+
+		var value string = ""
+		if len(keyAndValue) > 1 {
+			value = strings.TrimSpace(keyAndValue[1])
+		}
+
+		if isInstanceIdentifier(key) {
+			sqlInstance = value
+			if databaseName != emptyDatabaseName {
+				return
+			}
+		} else if isDatabaseIdentifier(key) {
+			databaseName = value
+			if sqlInstance != emptyServerName {
+				return
+			}
+		}
+	}
+
+	return
+}
+
+// parseConnectionStringURL parses a URL-formatted connection string and returns the SQL instance and database name
+func parseConnectionStringURL(connectionString string) (sqlInstance string, databaseName string) {
+	sqlInstance = emptyServerName
+	databaseName = emptyDatabaseName
+
+	u, err := url.Parse(connectionString)
+	if err != nil {
+		return
+	}
+
+	host, _, err := net.SplitHostPort(u.Host)
+	if err != nil {
+		// The host did not have a port
+		sqlInstance = u.Host
+	} else {
+		// The host did have a port (e.g. "the.host.com:1234")
+		sqlInstance = host
+	}
+
+	if len(u.Path) > 1 {
+		// There was a SQL instance name specified in addition to the host
+		// E.g. "the.host.com:1234/InstanceName" or "the.host.com/InstanceName"
+		sqlInstance = sqlInstance + "\\" + u.Path[1:]
```

Comment:
Do you really need backslashes here or can we do with slash?

Reply:
While rather annoying, the `\` characters are the official format for SQL Server named instances. See here:
- https://docs.microsoft.com/en-us/previous-versions/sql/sql-server-2005/ms165614(v=sql.90)#sql-server-express-default-and-named-instances
- https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/connection-string-syntax#connect-to-a-named-instance-of-sql-server


It's not a deal breaker to deviate from the format. But it is the format that users would expect to see.

Reply:
Ok then please keep it and close this as resolved.

Reply:
Oh one more question, if the path has more than one "segment" they are probably separated by `/` aren't they? Do you want to replace those with backslash also?

Reply:
Good point. It's unclear how to handle that situation. The SQL Server driver for Go just ignores any extra segments, so that's also what I did here. (see https://github.com/denisenkom/go-mssqldb/blob/master/conn_str.go#L313).